src/core/midi/MPEManager.cpp

```diff
@@ -0,0 +1,133 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "MPEManager.h"
+
+#include "MidiEventProcessor.h"
+#include "MidiPort.h"
+
+namespace lmms
+{
+
+int MPEManager::findAvailableChannel()
+{
+	// For the lower zone, the first channel, channel 0, is the Manager channel. The channels after that are Member channels.
+	// For the upper zone, the last channel, channel 15, is the Manager channel. The channels below 15 are Member channels.
+
+	const int numChannels = m_zone == MPEZone::Lower
+		? m_numChannelsLowerZone
+		: m_numChannelsUpperZone;
+	const int numMemberChannels = numChannels - 1;
+	// Page 23, B.6, if there are no member channels the zone should deactivate.
+	if (numMemberChannels <= 0) { return -1; }
+
+	const int firstMemberChannel = m_zone == MPEZone::Lower
+		? 1
+		: 15 - numMemberChannels;
+	const int lastMemberChannel = m_zone == MPEZone::Lower
+		? numMemberChannels
+		: 14;
+
+	// The MPE specification, page 27 Appendix E, does not prohibit NoteOn/NoteOff signals on the Manager channels.
+	// However, notes should ideally not be routed to Manager channels, since then you can't control their pitch individually
+	// without affecting the pitch of the whole instrument.
+
+	// Page 17, Appendix A.3 gives some ideas on how to implement a good note routing system which aims to minimize the
+	// chance for notes being incorrectly pitch bent due to being on the same channel as another note being pitch bent.
+
+	// Their proposed method routes notes to the channels which have the fewest active notes.
+	// If two channels have the same number of notes, the ones with the oldest NoteOff signal is preferred.
+
+	// Find member channel with fewest notes/oldest NoteOff signal
+	int bestChannel = firstMemberChannel;
+	int bestNoteCount = m_channelNoteCounts[bestChannel];
+	// Technically we are not using time, but instead the count of NoteOff events up to that point. This way it does not depend on the current time.
+	int bestNoteOffTime = m_channelNoteOffTimes[bestChannel];
+	for (int channel = firstMemberChannel; channel <= lastMemberChannel; ++channel)
+	{
+		if (m_channelNoteCounts[channel] < bestNoteCount
+			|| (m_channelNoteCounts[channel] == bestNoteCount && m_channelNoteOffTimes[channel] < bestNoteOffTime))
+		{
+			bestChannel = channel;
+			bestNoteCount = m_channelNoteCounts[channel];
+			bestNoteOffTime = m_channelNoteOffTimes[channel];
+		}
+	}
+	return bestChannel;
+}
+
+
+void MPEManager::sendMPEConfigSignals(MidiEventProcessor* proc)
```

Comment:
Since this cannot be null, you might want to add a null-pointer check in this method just to be sure a nullptr is never passed